FIX Fix `DecisionTree*` partitioning with missing values present

betatim opened the pull request:
Todo:
* [x] clean up commit
* [x] add a test that fails on `main` and passes on this PR

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
#32178

#### What does this implement/fix? Explain your changes.
This is a simple fix, but it took a while to find. The problem occurs only if you first look at a different feature and then consider the feature that has missing values. This explains (I think) why it happens roughly half the time when you run the following snippet:
```python
In [2]: import numpy as np
   ...: from sklearn.tree import DecisionTreeRegressor
   ...:
   ...: X = np.vstack([
   ...:     [0, 0, 0, 0, 1, 2, 3, 4],
   ...:     [1, 2, 1, 2, 1, 2, 1, 2]
   ...: ]).swapaxes(0, 1).astype(float)
   ...: y = [0, 0, 0, 0, 1, 1, 1, 1]
   ...: X[X==0] = np.nan
   ...: for i in range(50):
   ...:   tree = DecisionTreeRegressor(max_depth=1, random_state=i, min_impurity_decrease=0.002).fit(X, y)
   ...:   print(tree.tree_.impurity)
```

The way to tell that something went wrong is that the impurity is not `[0.25 0.   0.  ]` but `[0.25   0.1875 0.1875]`. If you uncomment the `print` statements (in the first commit of this PR) you will see the following printed when running the above snippet for a random state that is broken (`i=2` on my computer).
```
before partitioning best_n_missing=4 partition_end=3 start=0 end=7 best_threshold=inf
X[i, best_feature]=1.0 i=4
X[i, best_feature]=nan i=0
X[i, best_feature]=3.0 i=6
X[i, best_feature]=nan i=2
X[i, best_feature]=nan i=3
X[i, best_feature]=nan i=1
X[i, best_feature]=4.0 i=7
X[i, best_feature]=2.0 i=5
---- p=0 samples[p]=4 current_value=1.0 end=7
X[i, best_feature]=1.0 i=4
X[i, best_feature]=nan i=0
X[i, best_feature]=3.0 i=6
X[i, best_feature]=nan i=2
X[i, best_feature]=nan i=3
X[i, best_feature]=nan i=1
X[i, best_feature]=4.0 i=7
X[i, best_feature]=2.0 i=5
---- p=1 samples[p]=0 current_value=nan end=7
X[i, best_feature]=1.0 i=4
X[i, best_feature]=nan i=0
X[i, best_feature]=3.0 i=6
X[i, best_feature]=nan i=2
X[i, best_feature]=nan i=3
X[i, best_feature]=nan i=1
X[i, best_feature]=4.0 i=7
X[i, best_feature]=2.0 i=5
---- p=2 samples[p]=6 current_value=3.0 end=6
X[i, best_feature]=1.0 i=4
X[i, best_feature]=2.0 i=5
X[i, best_feature]=3.0 i=6
X[i, best_feature]=nan i=2
X[i, best_feature]=nan i=3
X[i, best_feature]=nan i=1
X[i, best_feature]=4.0 i=7
X[i, best_feature]=nan i=0
after partitioning
X[i, best_feature]=1.0 i=4
X[i, best_feature]=2.0 i=5
X[i, best_feature]=3.0 i=6
X[i, best_feature]=nan i=2
X[i, best_feature]=nan i=3
X[i, best_feature]=nan i=1
X[i, best_feature]=4.0 i=7
X[i, best_feature]=nan i=0
[0.25   0.1875 0.1875]
```

If you look at the last print out of the samples you see that `4.0` is in amongst the missing values. Counting how how many times we "swapped" values/where `p` ends is what makes me think we need to either use `<=` for the condition or `intp_t partition_end = end - best_n_missing + 1`

#### Any other comments?

This was an interesting puzzle. I used the snippet from the issue that @cakedev0 posted and then asked Cursor to explain to me how it could be that this happens. It fairly quickly figured out that it must be something to do with which feature is considered first (because we permute the feature order). It then hinted at a problem with samples not being sorted properly. But the fixes it suggested were to add a special case to `partition_samples_final` for when the threshold was `np.inf`. This kind of fixed the problem but not completely. It then suggested sorting the values before calling the function. At this point I decided that Cursor was doing what it often does: come up with "overfitted" solutions (specific solution for just the one problem you asked it about). Having had it point to sorting being a problem and that it only happens half the time I then went back to the old school "add print statements" approach :D Combined with a hunch that an off by one error is likely given how much gymnastics happens to calculate all the start and end positions.

Overall a fun experience. Thanks for making that initial snippet @cakedev0 it was a good minimal reproducer!

betatim committed:
Fix partitioning with missing values present

We need to examine one more value to make sure we look at all the non
missing values.

betatim commented:
Do you have a nice unittest for this issue? Fitting a bunch of trees and checking that all of them get the right `impurity`/have enough splits seems a bit of a round about way of testing this :-/

betatim committed:
Remove debugging code

betatim committed:
Add non regression test

cakedev0 commented on sklearn/tree/tests/test_tree.py:
```diff
@@ -2873,3 +2873,23 @@ def test_sort_log2_build():
+    # The important thing here is that we want to have two trees, where each one tries
+    # one of the two features first. The resulting tree should be the same in both
+    # cases. The way to control which feature is tried first is `random_state`.
+    tree1 = DecisionTreeRegressor(max_depth=1, random_state=1).fit(X, y)
+    tree2 = DecisionTreeRegressor(max_depth=1, random_state=2).fit(X, y)
```
You can duplicate the second feature many times (10 times for instance), this will highly increase the failure probability on main:

<details>

```Python
In [4]: import numpy as np
   ...: from sklearn.tree import DecisionTreeRegressor
   ...: 
   ...: X = np.vstack([
   ...:     [0, 0, 0, 0, 1, 2, 3, 4],
   ...:     *([[1, 2, 1, 2, 1, 2, 1, 2]]*10)
   ...: ]).swapaxes(0, 1).astype(float)
   ...: y = [0, 0, 0, 0, 1, 1, 1, 1]
   ...: X[X == 0] = np.nan
   ...: n_leaves_w_missing = []
   ...: for _ in range(1000):
   ...:     tree = DecisionTreeRegressor(max_depth=1, min_impurity_decrease=0.25).fit(X, y)
   ...:     n_leaves_w_missing.append(tree.tree_.n_leaves)
   ...: 
   ...: print(np.bincount(n_leaves_w_missing))
[  0 911  89]
```
</details>

Also, I would suggest using `random_state=global_random_seed` / `random_state=global_random_seed + 1` / ... so that if a regression happens, it's very likely that at least one of the CI pipeline will fail.

Catching a regression with probability 99.9% is really good enough IMO.

> this is where flaky tests come from

In that case, the check we do is supposed to pass for any random seed. So it's never going to become flaky.

Globally, this is a good toy-example test IMO

ogrisel commented:
> I personally love tests with many random inputs. Toy examples should only be used for debugging IMO. It's still good to have them as tests, as it makes it easy to debug regressions. But it's really not enough to build confidence.

Unfortunately, such randomized tests are costly to run so we need to strike a balance. Ideally, individual tests should last no longer than 1s on a usual CI runner.

betatim committed:
Whats new

betatim committed:
Repeatedly fit and check a tree

By repeating the fit and checking the computed impurity we should catch
the bug. The bug should happen in ~50% of fits, so ten fits are likely
to trigger it.

betatim committed:
Merge branch 'main' into fix-impurity-calculation-with-missing

cakedev0 reviewed (approved):
LGTM. 

Nit: I would rather go with 20 trees as 1/2^10 is still 0.1% but maybe I'm too cautious :sweat_smile:

betatim committed:
Twenty trees should do it

ogrisel commented on doc/whats_new/upcoming_changes/sklearn.tree/32351.fix.rst:
```diff
@@ -0,0 +1,3 @@
+- Fix decision tree splitting with missing values present in some features. In some cases the last
+  non-missing sample would not be partitioned correctly.
+  By :user:`Tim Head <betatim>`.
```
Let's credit Arthur for the initial analysis, minimal reproducer and concurrent fix in his other PR and contribution about how to write a proper non-regression test for this.

```suggestion
  By :user:`Tim Head <betatim>` and :user:`Arthur Lacote <cakedev0>`.
```

ogrisel reviewed (approved):
LGTM.

ogrisel committed:
Co-credit Arthur for this fix.

ogrisel committed:
Merge branch 'main' into fix-impurity-calculation-with-missing
